addOns/automation/src/main/java/org/zaproxy/addon/automation/AutomationEventPublisher.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2021 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.addon.automation;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.zaproxy.zap.ZAP;
+import org.zaproxy.zap.eventBus.Event;
+import org.zaproxy.zap.eventBus.EventPublisher;
+
+/** @since 2.8.0 */
+public class AutomationEventPublisher implements EventPublisher {
+
+    private static AutomationEventPublisher publisher = null;
+
+    /** Indicates that a plan has started to run. */
+    public static final String PLAN_STARTED = "plan.started";
+
+    public static final String PLAN_ENV_CREATED = "plan.env.created";
+
+    public static final String PLAN_FINISHED = "plan.finished";
+
+    public static final String PLAN_ERROR_MESSAGE = "plan.error";
+
+    public static final String PLAN_WARNING_MESSAGE = "plan.warning";
+
+    public static final String PLAN_INFO_MESSAGE = "plan.info";
+
+    public static final String JOB_STARTED = "job.started";
+
+    public static final String JOB_FINISHED = "job.finished";
+
+    public static final String JOB_ID = "jobId";
+    public static final String JOB_NAME = "jobName";
+    public static final String JOB_TYPE = "jobType";
+    public static final String PLAN_ID = "planId";
+    public static final String MESSAGE = "message";
+
+    @Override
+    public String getPublisherName() {
+        return AutomationEventPublisher.class.getCanonicalName();
+    }
+
+    public static synchronized AutomationEventPublisher getPublisher() {
+        if (publisher == null) {
+            publisher = new AutomationEventPublisher();
+            ZAP.getEventBus()
+                    .registerPublisher(
+                            publisher,
+                            PLAN_STARTED,
+                            PLAN_ENV_CREATED,
+                            PLAN_FINISHED,
+                            PLAN_ERROR_MESSAGE,
+                            PLAN_WARNING_MESSAGE,
+                            PLAN_INFO_MESSAGE,
+                            JOB_STARTED,
+                            JOB_FINISHED);
+        }
+        return publisher;
+    }
+
+    public static void publishMessageEvent(String event, String msg) {
+        Map<String, String> map = new HashMap<>();
+        map.put(MESSAGE, msg);
```

Comment:
We _can_ but does it really help? It just seems like it would be another thing to maintain with little advantage

Reply:
In case we are sure that the size is always one then the default size of 16 will increase the memory footprint and if this method is invoked multiple times we can avoid the memory thing. we can use collections.singletonMap also